lib/AppStore/Service.php

```
@@ -0,0 +1,150 @@
+		$info = $this->getInstalledAppInfo($ocsId);
+		if (is_null($info)) {
+			throw new \InvalidArgumentException("App ($ocsId) is not installed");
+		}
+		$marketInfo = $this->getAppInfo($ocsId);
+		if (is_null($marketInfo)) {
+			throw new \InvalidArgumentException("App ($ocsId) is not known at the marketplace.");
+		}
+		$marketVersion = (string) $marketInfo['version'];
+		$currentVersion = (string) $info['version'];
+		if (version_compare($marketVersion, $currentVersion, '>')) {
+			return $marketVersion;
+		}
+		return false;
+	}
+
+	private function getAppInfo($ocsId) {
+		//
+		// TODO: read from cached information as stored by using ./occ market:update
+		//
+		$ocsClient = new OCSClient(
+			\OC::$server->getHTTPClientService(),
+			\OC::$server->getConfig(),
+			\OC::$server->getLogger()
+		);
+		return $ocsClient->getApplication($ocsId, \OCP\Util::getVersion());
+	}
+
+	private function getInstalledAppInfo($ocsId) {
+		$apps = $this->appManager->getAllApps();
+		foreach ($apps as $app) {
+			$info = $this->appManager->getAppInfo($app);
+			if (isset($info['ocsid']) && (int)$info['ocsid'] === (int)$ocsId) {
+				return $info;
+			}
+		}
+
+		return null;
+	}
+
+	public function updateApp($ocsId) {
+		$info = $this->getInstalledAppInfo($ocsId);
+		if (is_null($info)) {
+			throw new \InvalidArgumentException('App is not installed');
+		}
+
+		// download package
+		$package = $this->downloadPackage($ocsId);
+		$this->appManager->updateApp($package);
+	}
+
+	public function getUpdates() {
+		$result = [];
+		$apps = $this->appManager->getAllApps();
+		foreach ($apps as $app) {
+			$info = $this->appManager->getAppInfo($app);
+			if (isset($info['ocsid'])) {
+				try {
+					$ocsId = $info['ocsid'];
+					$newVersion = $this->updateAvailable($ocsId);
+					if ($newVersion) {
+						$result[$app] = [
+							'version' => $newVersion,
+							'ocsid' => $ocsId
+						];
+					}
+				} catch (\InvalidArgumentException $ex) {
+				}
```

add PHPDoc that explains why ignoring this exception is intentional